esp/services/esdl_svc_engine/esdl_binding.cpp

```diff
@@ -889,49 +889,96 @@ void EsdlServiceImpl::handleFinalRequest(IEspContext &context,
                    "No target URL configured for %s!", mthdef.queryMethodName());
     }
 
-    try
-    {
-        Owned<IPTree> responseTree = createPTreeFromXMLString(out, ipt_ordered);
-        auto responseResult = responseTree->queryBranch("//Results/Result");
+    auto echoInsertionPoint = strstr(out, "</Result></Results>");
 
-        if (responseResult)
+    if (echoInsertionPoint)
+    {
+        try
         {
-            auto replaceTag = [](StringBuffer& xml, const char* tag, const char* var)
+            auto queryname = (isroxie ? tgtcfg->queryProp("@queryname") : nullptr);
+            auto requestname = mthdef.queryRequestType();
+            const char* qname = nullptr;
+            const char* content = nullptr;
+            StringBuffer encodedContent;
+            StringBuffer tmpOut;
+
+            XmlPullParser xpp(soapmsg.str(), soapmsg.length());
+            StartTag stag;
+            EndTag etag;
+            int type = XmlPullParser::END_DOCUMENT;
+
+            tmpOut.ensureCapacity(out.length() + soapmsg.length());
+            tmpOut.append(echoInsertionPoint - out.str(), out);
+            xpp.setSupportNamespaces(false); // parser throws exceptions when it encounters undefined namespace prefixes, such as CRT's 'xsdl'
+            while ((type = xpp.next()) != XmlPullParser::END_DOCUMENT)
             {
-                VStringBuffer replaceThis("<%s>", tag);
-                VStringBuffer withThat("<%s name=\"%s\">", var, tag);
-
-                xml.replaceString(replaceThis, withThat);
-
-                replaceThis.insert(1, '/');
-                withThat.setLength(strlen(var) + 1);
-                withThat.insert(1, '/');
-                withThat.append('>');
-
-                xml.replaceString(replaceThis, withThat);
-            };
-
-            if (isroxie && tgtcfg->hasProp("@queryname"))
-                replaceTag(soapmsg, tgtcfg->queryProp("@queryname"), "roxiequery");
-            replaceTag(soapmsg, mthdef.queryRequestType(), "esprequest");
-            soapmsg.replaceString("xsdl:", "xsdl_");
-
-            Owned<IPTree> requestTree = createPTreeFromXMLString(soapmsg, ipt_ordered);
-            auto bodyTree = requestTree->queryBranch("soap:Body");
+                switch (type)
+                {
+                case XmlPullParser::START_TAG:
+                    xpp.readStartTag(stag);
+                    qname = stag.getQName();
+                    if (strcmp(qname, "soap:Envelope") == 0)
+                        tmpOut << "<Dataset name=\"DesdlSoapRequestEcho\">";
+                    else if (strcmp(qname, "soap:Body") == 0)
+                        tmpOut << "<Row>";
+                    else if (queryname && strcmp(qname, queryname) == 0)
+                        tmpOut << "<roxierequest name=\"" << queryname << "\">";
```

Comment:
There may be a common routine already to serialize all the children beneath the current tag.  If not we should add one in the future.

Reply:
Unfortunately I'm modifying many children of the roxie request, so serializing all children at that level wouldn't work. If the custom request transform namespace issue gets resolved (there's an open issue about the namespace preventing use of the roxie test UI, and the logging code has the same issue), then we could benefit from optimization.

Reply:
ok.  Not sure where namespace you are talking about comes from, but it should either be removed, or properly defined at any layer that creates, uses, or passes it on.

Reply:
Every statement in the CRT includes `xsdl` as a namespace prefix. The binding requires the prefix to be present, without expecting a namespace declaration. Any pull parser supporting namespaces will throw an exception because of it.

I tried adding a namespace declaration to my transform, without success. I might have done it incorrectly, but I had to delete and recreate the binding after the attempt and didn't have time to keep fighting with it. That's when I opened https://track.hpccsystems.com/browse/HPCC-20883.